core/services/vrf/listener_v2_types.go

```diff
@@ -0,0 +1,202 @@
+	lbs           []log.Broadcast
+	maxLinks      []interface{}
+}
+
+func newBatchFulfillment(result vrfPipelineResult) *batchFulfillment {
+	return &batchFulfillment{
+		proofs: []batch_vrf_coordinator_v2.VRFTypesProof{
+			batch_vrf_coordinator_v2.VRFTypesProof(result.proof),
+		},
+		commitments: []batch_vrf_coordinator_v2.VRFTypesRequestCommitment{
+			batch_vrf_coordinator_v2.VRFTypesRequestCommitment(result.reqCommitment),
+		},
+		totalGasLimit: result.gasLimit,
+		runs: []*pipeline.Run{
+			&result.run,
+		},
+		reqIDs: []*big.Int{
+			result.req.req.RequestId,
+		},
+		lbs: []log.Broadcast{
+			result.req.lb,
+		},
+		maxLinks: []interface{}{
+			result.maxLink,
+		},
+	}
+}
+
+type batchFulfillments struct {
+	fulfillments  []*batchFulfillment
+	batchGasLimit uint64
+	currIndex     int
+}
+
+func newBatchFulfillments(batchGasLimit uint64) *batchFulfillments {
+	return &batchFulfillments{
+		fulfillments:  []*batchFulfillment{},
+		batchGasLimit: batchGasLimit,
+		currIndex:     0,
+	}
+}
+
+func (b *batchFulfillments) addRun(result vrfPipelineResult) {
+	if len(b.fulfillments) == 0 {
+		b.fulfillments = append(b.fulfillments, newBatchFulfillment(result))
+	} else {
+		currBatch := b.fulfillments[b.currIndex]
+		if (currBatch.totalGasLimit + result.gasLimit) >= b.batchGasLimit {
+			// don't add to curr batch, add new batch and increment index
+			b.fulfillments = append(b.fulfillments, newBatchFulfillment(result))
+			b.currIndex++
+		} else {
+			// we're okay on gas, add to current batch
+			currBatch.proofs = append(currBatch.proofs, batch_vrf_coordinator_v2.VRFTypesProof(result.proof))
+			currBatch.commitments = append(currBatch.commitments, batch_vrf_coordinator_v2.VRFTypesRequestCommitment(result.reqCommitment))
+			currBatch.totalGasLimit += result.gasLimit
+			currBatch.runs = append(currBatch.runs, &result.run)
+			currBatch.reqIDs = append(currBatch.reqIDs, result.req.req.RequestId)
+			currBatch.lbs = append(currBatch.lbs, result.req.lb)
+			currBatch.maxLinks = append(currBatch.maxLinks, result.maxLink)
+		}
+	}
+}
+
+func (lsn *listenerV2) processBatch(
+	l logger.Logger,
+	subID uint64,
+	fromAddress common.Address,
+	startBalanceNoReserveLink *big.Int,
+	batch *batchFulfillment,
+) (processedRequestIDs []string) {
+	start := time.Now()
+
+	// Enqueue a single batch tx for requests that we're able to fulfill based on whether
+	// they passed simulation or not.
+	payload, err := batchCoordinatorV2ABI.Pack("fulfillRandomWords", batch.proofs, batch.commitments)
+	if err != nil {
+		// should never happen
+		l.Errorw("Failed to pack batch fulfillRandomWords payload",
+			"err", err, "proofs", batch.proofs, "commitments", batch.commitments)
+		return
+	}
+
+	// Bump the total gas limit by a bit so that we account for the overhead of the batch
+	// contract's calling.
+	totalGasLimitBumped := uint64(float64(batch.totalGasLimit) * 1.15)
+	ll := l.With("numRequestsInBatch", len(batch.reqIDs),
+		"requestIDs", batch.reqIDs,
```

Comment:
Does this print out nicely? Want to make sure it's still easy to search by request ID

Reply:
The log ends up looking something like this:

```
2022-03-24T17:12:53.141+0200 [INFO]  Enqueueing batch fulfillment                                               coordinatorAddress=0xC80a24d6Ee900E8037Dd33A99F20575ea3582575 externalJobID=238489e7-148d-4120-acb3-d67cf62908c6 jobID=1 logger=unset@unset.TestVRFV2Integration_SingleConsumer_HappyPath_BatchFulfillment.VRFListenerV2 reqs=4 requestIDs=[52107590518753990297551874331065558019927806985271235729310933471538889517978,37661170113009891713477136563562430452544262167203968038611291464662037250977,25129323679989917518867569199072155508204031397994782432045688005979864206468] startBalance=5000000000000000000 startBalanceNoReservedLink=5000000000000000000 subID=1 
```

I think we can still search for this using Loki's `|~` operator, passing in the specific request ID we're concerned about.